### PR TITLE
[Draft] vm: add Expunged as final VM state and enable expunge VMs in Expunging state

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

main changes of this PR

- Introduce a new state Expunged which is final vm state (currently the final vm state is Expunging)
- Enable expunge a vm in Expunging state (current the action is disabled)
- Update vm state to Expunged if succeed to expunge a VM
- Update vm state to Error if fail to expunge a VM
- Add Expunged but not removed VMs to VM cleanup task.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test matrix

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@DaanHoogland 
could you please review this PR ? thanks

File: api/src/main/java/com/cloud/vm/VirtualMachine.java [hunk 1]
```diff
@@ -122,7 +123,8 @@ public static StateMachine2<State, VirtualMachine.Event, VirtualMachine> getStat
             s_fsm.addTransition(new Transition<State, Event>(State.Stopping, VirtualMachine.Event.AgentReportStopped, State.Stopped, Arrays.asList(new Impact[]{Impact.USAGE})));
             s_fsm.addTransition(new Transition<State, Event>(State.Stopping, VirtualMachine.Event.StopRequested, State.Stopping, null));
             s_fsm.addTransition(new Transition<State, Event>(State.Stopping, VirtualMachine.Event.AgentReportShutdowned, State.Stopped, Arrays.asList(new Impact[]{Impact.USAGE})));
-            s_fsm.addTransition(new Transition<State, Event>(State.Expunging, VirtualMachine.Event.OperationFailed, State.Expunging,null));
+            s_fsm.addTransition(new Transition<State, Event>(State.Expunging, VirtualMachine.Event.OperationSucceeded, State.Expunged, null));
+            s_fsm.addTransition(new Transition<State, Event>(State.Expunging, VirtualMachine.Event.OperationFailed, State.Error,null));
```
Comment (DaanHoogland) [inline api/src/main/java/com/cloud/vm/VirtualMachine.java:127]: sonar comment: 
```suggestion
            s_fsm.addTransition(new Transition<>(State.Expunging, VirtualMachine.Event.OperationSucceeded, State.Expunged, null));
            s_fsm.addTransition(new Transition<>(State.Expunging, VirtualMachine.Event.OperationFailed, State.Error,null));
```
makes sense not to adhere to this old standard.

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by weizhouapache** [comment]
@DaanHoogland 
moved to draft. let's revisit afterwards

**Comment by weizhouapache** [comment]
small fix: https://github.com/apache/cloudstack/pull/7086

**Comment by rohityadavcloud** [comment]
@weizhouapache should this be closed as the simpler fix is merged now? https://github.com/apache/cloudstack/pull/7086/files ?

**Pull request closed without merging**